File: leetcode/19.py
class Solution:
  def removeNthFromEnd(self, head:Optional[ListNode], n:int) -> Optional[ListNode]:
    l, i = 0, 0
    ptr1 = head
    while head:
      l += 1
      head = head.next
    if n == l:
      return ptr1.next
    ptr2 = ptr1
    while ptr1:
      if l-i == n+1:
        ptr1.next = ptr1.next.next
        break
      i += 1
      ptr1 = ptr1.next
    return ptr2

File: leetcode/19.cpp
class Solution {
public:
  ListNode* removeNthFromEnd(ListNode* head, int n) {
    int l = 0, i = 0;
    ListNode* ptr1 = head;
    while (head != nullptr) {
      l++;
      head = head->next;
    }
    if (n == l) return ptr1->next;
    ListNode* ptr2 = ptr1;
    while (ptr1 != nullptr) {
      if (l-i-1 == n) {
        ListNode* temp = ptr1->next;
        ptr1->next = temp->next;
        delete temp;
        break;
      }
      i++;
      ptr1 = ptr1->next;
    }
    return ptr2;
  }
};

